src/Jackalope/Version/VersionHandler.php

```diff
@@ -235,15 +236,203 @@ private function createFrozenNode(NodeInterface $versionNode, NodeInterface $nod
+        // also see (https://github.com/jackalope/jackalope/issues/247)
+
+        if ($frozenNode->hasProperty('jcr:frozenMixinTypes')) {
+            $node->setProperty('jcr:mixinTypes', $frozenNode->getPropertyValue('jcr:frozenMixinTypes'));
+        } elseif ($node->hasProperty('jcr:mixinTypes')) {
+            $node->getProperty('jcr:mixinTypes')->remove();
+        }
+
+        // handle properties present on the frozen node
+        foreach ($frozenNode->getProperties() as $property) {
+            /** @var PropertyInterface $property */
+            $propertyName = $property->getName();
+            if ($propertyName == 'jcr:frozenPrimaryType'
+                || $propertyName == 'jcr:frozenMixinTypes'
+                || $propertyName == 'jcr:frozenUuid'
+            ) {
+                continue;
+            }
+
+            if ($node->hasProperty($propertyName)) {
+                $nodeProperty = $node->getProperty($propertyName);
+                $onParentVersion = $nodeProperty->getDefinition()->getOnParentVersion();
+                if ($onParentVersion == OnParentVersionAction::COPY || $onParentVersion == OnParentVersionAction::VERSION) {
+                    $nodeProperty->setValue($property->getValue());
+                }
+            } else {
+                // cannot check the onParentVersion attribute of a non existing property
+                // but if the property exists on the frozen node it has to be copy or version
+                $node->setProperty($propertyName, $property->getValue());
+            }
+        }
+
+        // handle properties present on the node but not on the frozen node
+        foreach ($node->getProperties() as $propertyName => $property) {
+            if ($frozenNode->hasProperty($propertyName)) {
+                continue;
+            }
+
+            $onParentVersion = $property->getDefinition()->getOnParentVersion();
+
+            switch ($onParentVersion) {
+                case OnParentVersionAction::COPY:
+                case OnParentVersionAction::VERSION:
+                case OnParentVersionAction::ABORT:
+                    $property->remove();
+                    break;
+                case OnParentVersionAction::INITIALIZE:
+                case OnParentVersionAction::COMPUTE: // implementation-specific, decided to handle as INITIALIZE
+                    // TODO how to get default value of property?
+            }
+        }
+
+        // TODO handle identifier collisions
+
+        // handle child nodes present on the frozen node
+        foreach ($frozenNode->getNodes() as $frozenChildNode) {
+            /** @var NodeInterface $frozenChildNode */
+            if (!$removeExisting) {
+                // TODO check occurence of node in repository
+                throw new NotImplementedException('Check for $removeExisting not implemented yet');
+            }
+
+            // TODO handle different onParentVersion values
```

Comment:
does the spec say anything about node types of children of that frozenNode? maybe the whole subtree should be frozen? or the children would retain their original node type.

either way, getNodeType and getDefinition have to be consistent, i am surprised this is not the case. smells like we have code duplication between getDefinition and getNodeType... do you have the java jackrabbit client running? then you could try to recreate the situation and see what jackrabbit is doing. either the children just keep their node types, or they are frozen too and their original node type is in a separate property, just like for the main frozen node. (the more i think about this, the more sense it would make to freeze the whole subtree. otherwise you could change a child of a frozen node which makes no sense)